File: java/Medium/_948_BagOfTokens.java
package Medium;

import java.util.*;

import org.testng.annotations.Test;

public class _948_BagOfTokens {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int l = 0, r = tokens.length-1,score = 0, points = 0;
        while(l<=r ){
            if(power >= tokens[l]){
                power -= tokens[l++];
                score = Math.max(score,++points);
            }else if(points > 0){
                points--;
                power += tokens[r--];
            }else{
                break;
            }
        }
        return score;
    }


    @Test
    public void test(){
        System.out.println(bagOfTokensScore(new int[]{100,200}, 150));
        System.out.println(bagOfTokensScore(new int[]{100,200,300,400}, 200));
        System.out.println(bagOfTokensScore(new int[]{100}, 50));
    }
}

File: Python3/_948_BagOfTokens.py
import math
from typing import List


class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        res = 0
        l = 0
        r = len(tokens)-1
        points = 0

        while l <= r :
            if power >= tokens[l]:
                power -= tokens[l]
                points += 1
                res = max(res,points)
                l+=1
            elif points > 0:
                power += tokens[r]
                r-=1
                points -=1
            else:
                break



        return res